### PR TITLE
Move bs_worker_thread from main repo

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Move bs_worker_thread from main repo

### CODE REVIEW

File: src/worker_thread.c [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+                    if (worker_thread->execute_worker_event == NULL)
+                    {
+                        /* Codes_SRS_WORKER_THREAD_01_008: [ If any error occurs, worker_thread_create shall fail and return NULL. ]*/
+                        LogError("Cannot create process event");
+                    }
+                    else
+                    {
+                        goto all_ok;
+
+                        //(void)CloseHandle(worker_thread->execute_worker_event);
```
Comment (jebrando) [inline src/worker_thread.c:123]: >//(void)CloseHandle(worker_thread->execute_worker_event); [](http://example.com/codeflow?start=24&length=57)

Can you please remove the comment

File: CMakeLists.txt [hunk 2]
```diff
@@ -192,6 +192,19 @@ set(c_util_h_files
...
+#add worker_thread only on Windows (until we fix it)
+set(c_util_c_files
+    ${c_util_c_files}
+    ./src/worker_thread.c
+)
+
+set(c_util_h_files
+    ${c_util_h_files}
+    ./inc/c_util/worker_thread.h
+)
```
Comment (jebrando) [inline CMakeLists.txt:205]: nitpick: can you indent between the if()...endif().  Keep the cmakelist pretty.

File: devdoc/worker_thread_requirements.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
+﻿`worker_thread` requirements
+================
+
+## Overview
+
+`worker_thread` is a helper module that implements execution of a work item on a separate thread. It can be used in batching operations, sequencing completions, etc.
+The user instantiates a BS worker thread instance and then signals when the work shall be executed by calling `worker_thread_schedule_process`.
```
Comment (jebrando) [inline devdoc/worker_thread_requirements.md:7]: >BS [](http://example.com/codeflow?start=24&length=2)

Can you please remove the BS here and just have it as:

"a worker thread ..."

**Review by jebrando — APPROVED** [review]
:shipit:

File: devdoc/worker_thread_requirements.md (hunk 3 above)
Comment (dcristoloveanu) [reply devdoc/worker_thread_requirements.md:7]: Removed BS

File: devdoc/worker_thread_requirements.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+Note: The BS worker thread module was created in order to ease testing of the user components.
```
[the remark this answers is not in the document]
Comment (dcristoloveanu) [reply devdoc/worker_thread_requirements.md:9]: Removed more BS

File: src/worker_thread.c (hunk 1 above)
Comment (dcristoloveanu) [reply src/worker_thread.c:123]: That one is left on purpose. This is the only exception we have for leaving comments in the code (it is cleanup that would be enabled at the moment you add another if in the function).

File: CMakeLists.txt (hunk 2 above)
Comment (dcristoloveanu) [reply CMakeLists.txt:205]: It's very pretty now :-)

File: CMakeLists.txt [hunk 5]
```diff
@@ -192,6 +192,19 @@ set(c_util_h_files
     ./inc/c_util/uuid_string.h
 )
 
+if(WIN32)
+    #add worker_thread only on Windows (until we fix it)
```
Comment (mattdurak) [inline CMakeLists.txt:196]: >(until we fix it) [](http://example.com/codeflow?start=39&length=17)

Do we have a task to extract the Win32 APIs we use out to the PAL (or change the implementation to use something already in the PAL)?

File: devdoc/worker_thread_requirements.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
+﻿`worker_thread` requirements
+================
```
Comment (mattdurak) [inline devdoc/worker_thread_requirements.md:2]: nit: `#` header instead of `====`

File: CMakeLists.txt (hunk 5 above)
Comment (dcristoloveanu) [reply CMakeLists.txt:196]: ID Work Item Type Title Iteration Path State Assigned To Original Estimate
14543482 Task Have worker_thread be OS agnostic One\Custom\Azure Messaging Store\Iteration 261 To Do

File: devdoc/worker_thread_requirements.md (hunk 6 above)
Comment (dcristoloveanu) [reply devdoc/worker_thread_requirements.md:2]: Fixed.

File: devdoc/worker_thread_requirements.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+## Overview
+
+`worker_thread` is a helper module that implements execution of a work item on a separate thread. It can be used in batching operations, sequencing completions, etc.
```
[the remark this answers is not in the document]
Comment (dcristoloveanu) [reply devdoc/worker_thread_requirements.md:6]: Added single.

File: inc/c_util/worker_thread.h [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+#ifndef WORKER_THREAD_H
+#define WORKER_THREAD_H
+
+#include "umock_c/umock_c_prod.h"
+#ifdef __cplusplus
+extern "C" {
+#endif 
+
+    /* this is the callback that is to be called each time a batch process is scheduled */
```
Comment (mattdurak) [inline inc/c_util/worker_thread.h:12]: >each time [](http://example.com/codeflow?start=49&length=9)

nit: If the batch process is scheduled twice, the callback may only be called once. This comment implies a one-to-one mapping of callbacks and schedule_process calls.

File: tests/worker_thread_ut/worker_thread_ut.c [hunk 9]
```diff
@@ -0,0 +1,828 @@
...
+        HANDLE, hObject)
+        BOOL my_result = SetEvent(hObject);
+    MOCK_FUNCTION_END(my_result);
+
+    MOCK_FUNCTION_WITH_CODE(, DWORD, mocked_WaitForMultipleObjects,
+        DWORD, nCount,
+        CONST HANDLE*, lpHandles,
+        BOOL, bWaitAll,
+        DWORD, dwMilliseconds)
+        MOCK_FUNCTION_END(WAIT_OBJECT_0 + 1);
```
Comment (mattdurak) [inline tests/worker_thread_ut/worker_thread_ut.c:100]: nit: extra indent

File: tests/worker_thread_ut/worker_thread_ut.c [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+     real_gballoc_ll_free(ptr);
+}
+
+#include "testrunnerswitcher.h"
+#include "umock_c/umock_c.h"
+#include "umock_c/umocktypes_charptr.h"
+#include "umock_c/umocktypes_bool.h"
+#include "umock_c/umock_c_negative_tests.h"
+
+#include "c_pal/interlocked.h" /*included for mocking reasons - it will prohibit creation of mocks belonging to interlocked.h - at the moment verified through int tests - this is porting legacy code, temporary solution*/
```
[the remark this answers is not in the document]
Comment (dcristoloveanu) [reply tests/worker_thread_ut/worker_thread_ut.c:29]: Apparently we don't, removed it.

File: tests/worker_thread_ut/worker_thread_ut.c (hunk 9 above)
Comment (dcristoloveanu) [reply tests/worker_thread_ut/worker_thread_ut.c:100]: Fixed.

File: inc/c_util/worker_thread.h (hunk 8 above)
Comment (dcristoloveanu) [reply inc/c_util/worker_thread.h]: Added more explanation for this

**Pull request merged**